metabase/metabase
# Cleanup QuestionDataSource and fix breadcrumb links issues

This PR refactors the `QuestionDataSource` component, fixes #12616, fixes #17961, and also increases the Simple Mode font size (was supposed to be shipped with #18903, but I broke it at some point). `QuestionDataSource` is used in the query builder header and displays breadcrumb links to question data sources (database, schema, table, join tables, source question, etc.)

### To Verify

**Breadcrumb path with multiple joined tables only has a single link to the first table #17961**

1. Ask a question > Custom Question > Sample Dataset > Orders
2. Join Products and People tables
3. Save the question, exit the notebook editor
4. You should see the "Orders + People + Products" label in the header breadcrumbs
5. Click every table name, it should open the clicked table

**Breadcrumb links not working correctly in Nested Questions #12616**

1. Create a question on a schema-based database (can be Simple or Native) - save as "Q1"
2. Simple question > Saved Questions > "Q1" - save as "Q2"
3. You should not see Q1's collection name in data source breadcrumbs, it should be "DB / Q1"
4. Click "Q1", you should be navigated to `/question/:id-q1` instead of an ad-hoc query based on Q1

### Demo

<details>
<summary>Breadcrumb path with multiple joined tables only has a single link to the first table</summary>

**Before**

https://user-images.githubusercontent.com/17258145/141514930-cfe9a14c-d2f1-4511-83cb-3221e330b68a.mov

**After**

https://user-images.githubusercontent.com/17258145/141514946-574ec03b-78db-4553-bc00-84aac74de7b5.mov

</details>

<details>
<summary>Breadcrumb links not working correctly in Nested Questions</summary>

**Before**

https://user-images.githubusercontent.com/17258145/141514988-17d409d8-d138-4615-90e8-7c10bfbb92ac.mov

**After**

https://user-images.githubusercontent.com/17258145/141514993-34633337-1981-46ce-9c66-b251221b3327.mov

</details>
